src/articles/2024-01-21_DDEV-for-js.md

```diff
@@ -0,0 +1,327 @@
+
+### Using PM2
+
+[PM2](https://pm2.keymetrics.io/) is an "Advanced, production process manager for Node.JS". We can use this tool to start our app without clogging up an entire terminal pane and it will also allow us to monitor the process's logs, memory usage, and some other key stats. There are a ton of options available, but we'll just use a few basics.
+
+PM2's instructions tell us to install the tool globally, so we'll handle during DDEV's build process with a custom Dockerfile in `.ddev/web-build/Dockerfile`. For more info on how DDEV handles custom Dockerfiles, see [here](https://ddev.readthedocs.io/en/stable/users/extend/customizing-images/#adding-extra-dockerfiles-for-webimage-and-dbimage)
+
+```Dockerfile
+RUN npm install -g pm2
+```
+
+We can then setup our PM2 config by adding a file `apps.config.js`. This file will name our processes, tell PM2 where to run the start command, as well as what command to run. For a full list of config options see the [relevant docs](https://pm2.keymetrics.io/docs/usage/application-declaration/)
+
+```js
+module.exports = {
+  apps : [
+    {
+      name: "Svelte",
+      cwd: "/var/www/html/svelte",
+      script: "npm run dev"
+    },
+    {
+      name: "Keystone",
+      cwd: "/var/www/html/keystone",
+      script: "npm run dev"
+    }
+  ]
+}
+```
+
+### Post-start hooks
+
+With PM2 successfully added into the container & configured for our apps, we can now utilize [DDEV's Hooks](https://ddev.readthedocs.io/en/stable/users/configuration/hooks/) to run a command on startup. In our `.ddev/config.yml` file, add the following:
+
+```yml
+hooks:
+  post-start:
+    - exec: "pm2 start apps.config.js"
+```
+
+### Custom DDEV commands
+
+Finally, to interact with PM2 running in the container, we can add a [custom ddev command](https://ddev.readthedocs.io/en/stable/users/extend/custom-commands/) in `.ddev/commands/web/pm2`
+
+```bash
+#!/usr/bin/env bash
+
+## Description: Run PM2 commands in the container
+## Usage: pm2
+## Example: "ddev pm2 monit"
+
+pm2 "$@"
+```
+
+## One thing that still bugs me
+
+There's one part of this I still haven't quite figured out. Since I'm working on a mac with the new ARM architecture and the container is running linux, I have to run `npm install` in the container to get the right versions of some of the dependencies. If I install the requisite `node_module` on the host machine, the processes will error out if I try to run them inside the container. For right now the easiest way to handle this issue is to chain the install and start commands together in the PM2 config file:
+
+```js
+module.exports = {
+  apps : [
+    {
+      name: "Svelte",
+      cwd: "/var/www/html/svelte",
+      script: "rm -rf node_modules && npm install && npm run dev"
```

Comment:
Ah! You almost have it here. I actually think this is a decent way to run this. However, this is where I think a process manager like pm2 falls short. You often would want to restart without running npm ci.

Perhaps you could make it so npm ci is run once, on container startup, and not in pm2.

```suggestion
      script: "npm ci && npm run dev"
```

Reply:
@deviantintegral what about this:

```
{
      name: "Svelte",
      cwd: "/var/www/html/svelte",
      script: "npm run dev || (npm ci && npm run dev)"
}
```

In this way, the application attempts to go straight to start up, and in the case of a failure, does a `npm ci` and then trying again. Anything beyond that then becomes a debugging issue.

Reply:
I had to bail on `npm ci` for this blog post, thanks to how the Keystone quickstart works. It installs the app - without adding a package-lock.json for some reason. This absence breaks `npm ci`. I've updated the command to `npm run dev || (rm -rf node_modules && npm install && npm run dev)`. This should be able to handle host/container mismatch problems gracefully enough, I think.

Reply:
Makes sense. I think you should probably call out `npm ci` as the most reliable and preferred way to install dependencies, even if keystone can't do it.